File: crabzilla-vavr-example1/crabzilla-vavr-example1-service/src/main/java/io/github/crabzilla/example1/Example1EventProjector.java
package io.github.crabzilla.example1;

import io.github.crabzilla.core.DomainEvent;
import io.github.crabzilla.example1.readmodel.CustomerSummary;
import io.github.crabzilla.vertx.projection.EventProjector;
import lombok.extern.slf4j.Slf4j;
import org.jdbi.v3.core.Jdbi;

import static io.github.crabzilla.example1.customer.CustomerData.*;
import static io.vavr.API.*;
import static io.vavr.Predicates.instanceOf;

// tag::projector[]
@Slf4j
public class Example1EventProjector extends EventProjector<CustomerSummaryDao> {

  Example1EventProjector(String eventsChannelId, Class<CustomerSummaryDao> daoClass, Jdbi jdbi) {
    super(eventsChannelId, daoClass, jdbi);
  }

  @Override
  public void write(CustomerSummaryDao dao, String targetId, DomainEvent event) {

    log.info("writing event {} from channel {}", event, eventsChannelId);

    Match(event).of(
      Case($(instanceOf(CustomerCreated.class)), e ->
        run(() -> dao.insert(
                new CustomerSummary(e.getId().stringValue(), e.getName(), false)))),
      Case($(instanceOf(CustomerActivated.class)), e ->
        run(() -> dao.updateStatus(targetId, true))),
      Case($(instanceOf(CustomerDeactivated.class)), e ->
        run(() -> dao.updateStatus(targetId, false))),
      Case($(), o -> run(() -> log.warn("{} does not have any event projection handler", event))));
  }

}
// end::projector[]

File: crabzilla-kotlin-example1/crabzilla-kotlin-example1-service/src/main/java/io/github/crabzilla/example1/Example1EventProjector.kt
package io.github.crabzilla.example1


import io.github.crabzilla.core.DomainEvent
import io.github.crabzilla.example1.customer.CustomerActivated
import io.github.crabzilla.example1.customer.CustomerCreated
import io.github.crabzilla.example1.customer.CustomerDeactivated
import io.github.crabzilla.vertx.projection.EventProjector
import org.jdbi.v3.core.Jdbi
import org.slf4j.LoggerFactory

// tag::projector[]
class Example1EventProjector(channelId: String, daoClazz: Class<CustomerSummaryDao>, jdbi: Jdbi)

  : EventProjector<CustomerSummaryDao>(channelId, daoClazz, jdbi) {

  val log = LoggerFactory.getLogger(Example1EventProjector::class.java.simpleName)

  override fun write(dao: CustomerSummaryDao, targetId: String, event: DomainEvent) {

    log.info("event {} from channel {}", event, eventsChannelId)

    when (event) {
      is CustomerCreated -> dao.insert(CustomerSummary(targetId, event.name, false))
      is CustomerActivated -> dao.updateStatus(targetId, true)
      is CustomerDeactivated -> dao.updateStatus(targetId, false)
      else -> println("${event.javaClass.simpleName} does not have any event projection handler")
    }
  }
}
// end::projector[]
